File: ChronobaseCS/src/net/verdagon/chronobasecs/generator/chronobase/RootIncarnation.scala
package net.verdagon.chronobasecs.generator.chronobase

import net.verdagon.chronobasecs.compiled.{MutableS, SuperstructureS}

object RootIncarnation {

  def generateRootIncarnation(opt: ChronobaseOptions, ss: SuperstructureS): String = {

    val types: List[String] =
      ss.structs.filter(_.mutability == MutableS).map(_.name) ++
      ss.lists.filter(_.mutability == MutableS).map(_.name) ++
      ss.sets.filter(_.mutability == MutableS).map(_.name) ++
      ss.maps.filter(_.mutability == MutableS).map(_.name)

    s"public class RootIncarnation {\n" +
      s"  public readonly int version;\n" +
      s"  public int nextId;\n" +
      s"  public int hash;\n" +
      types
        .map(tyype => s"  public readonly SortedDictionary<int, VersionAndIncarnation<${tyype}Incarnation>> incarnations${tyype};\n")
        .mkString("") +
      s"  public RootIncarnation(int version, int nextId, int hash) {\n" +
      s"    this.version = version;\n" +
      s"    this.nextId = nextId;\n" +
      s"    this.hash = hash;\n" +
      types
        .map(tyype => s"    this.incarnations${tyype} = new SortedDictionary<int, VersionAndIncarnation<${tyype}Incarnation>>();\n")
        .mkString("") +
      s"  }\n" +
      s"  public RootIncarnation(\n" +
      s"      int newVersion,\n" +
      s"      int newNextId,\n" +
      s"      int newHash,\n" +
      s"      RootIncarnation that) {\n" +
      s"    this.version = newVersion;\n" +
      s"    this.nextId = newNextId;\n" +
      s"    this.hash = newHash;\n" +
      types
        .map(tyype => s"    this.incarnations${tyype} = new SortedDictionary<int, VersionAndIncarnation<${tyype}Incarnation>>(that.incarnations${tyype});\n")
        .mkString("") +
      s"  }\n" +
      s"}\n"
  }

}

File: Atharia/Geomancer/Model/RootIncarnation.cs
using System;
using System.Collections;

using System.Collections.Generic;

namespace Geomancer.Model {
public class RootIncarnation {
  public readonly int version;
  public int nextId;
  public int hash;
  public readonly SortedDictionary<int, VersionAndIncarnation<TerrainTileIncarnation>> incarnationsTerrainTile;
  public readonly SortedDictionary<int, VersionAndIncarnation<TerrainIncarnation>> incarnationsTerrain;
  public readonly SortedDictionary<int, VersionAndIncarnation<LevelIncarnation>> incarnationsLevel;
  public readonly SortedDictionary<int, VersionAndIncarnation<RandIncarnation>> incarnationsRand;
  public readonly SortedDictionary<int, VersionAndIncarnation<StrMutListIncarnation>> incarnationsStrMutList;
  public readonly SortedDictionary<int, VersionAndIncarnation<TerrainTileByLocationMutMapIncarnation>> incarnationsTerrainTileByLocationMutMap;
  public RootIncarnation(int version, int nextId, int hash) {
    this.version = version;
    this.nextId = nextId;
    this.hash = hash;
    this.incarnationsTerrainTile = new SortedDictionary<int, VersionAndIncarnation<TerrainTileIncarnation>>();
    this.incarnationsTerrain = new SortedDictionary<int, VersionAndIncarnation<TerrainIncarnation>>();
    this.incarnationsLevel = new SortedDictionary<int, VersionAndIncarnation<LevelIncarnation>>();
    this.incarnationsRand = new SortedDictionary<int, VersionAndIncarnation<RandIncarnation>>();
    this.incarnationsStrMutList = new SortedDictionary<int, VersionAndIncarnation<StrMutListIncarnation>>();
    this.incarnationsTerrainTileByLocationMutMap = new SortedDictionary<int, VersionAndIncarnation<TerrainTileByLocationMutMapIncarnation>>();
  }
  public RootIncarnation(
      int newVersion,
      int newNextId,
      int newHash,
      RootIncarnation that) {
    this.version = newVersion;
    this.nextId = newNextId;
    this.hash = newHash;
    this.incarnationsTerrainTile = new SortedDictionary<int, VersionAndIncarnation<TerrainTileIncarnation>>(that.incarnationsTerrainTile);
    this.incarnationsTerrain = new SortedDictionary<int, VersionAndIncarnation<TerrainIncarnation>>(that.incarnationsTerrain);
    this.incarnationsLevel = new SortedDictionary<int, VersionAndIncarnation<LevelIncarnation>>(that.incarnationsLevel);
    this.incarnationsRand = new SortedDictionary<int, VersionAndIncarnation<RandIncarnation>>(that.incarnationsRand);
    this.incarnationsStrMutList = new SortedDictionary<int, VersionAndIncarnation<StrMutListIncarnation>>(that.incarnationsStrMutList);
    this.incarnationsTerrainTileByLocationMutMap = new SortedDictionary<int, VersionAndIncarnation<TerrainTileByLocationMutMapIncarnation>>(that.incarnationsTerrainTileByLocationMutMap);
  }
}

}
